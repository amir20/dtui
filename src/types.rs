use tokio::sync::mpsc;

use crate::logs::LogEntry;

/// Host identifier for tracking which Docker host a container belongs to
pub type HostId = String;

/// Container metadata (static information)
#[derive(Clone, Debug)]
pub struct Container {
    pub id: String,
    pub name: String,
    pub status: String,
    pub stats: ContainerStats,
    pub host_id: HostId,
}

/// Container runtime statistics (updated frequently)
#[derive(Clone, Debug, Default)]
pub struct ContainerStats {
    pub cpu: f64,
    pub memory: f64,
    /// Network transmit rate in bytes per second
    pub network_tx_bytes_per_sec: f64,
    /// Network receive rate in bytes per second
    pub network_rx_bytes_per_sec: f64,
}

/// Unique key for identifying containers across multiple hosts
#[derive(Clone, Debug, Hash, Eq, PartialEq)]
pub struct ContainerKey {
    pub host_id: HostId,
    pub container_id: String,
}

impl ContainerKey {
    pub fn new(host_id: HostId, container_id: String) -> Self {
        Self {
            host_id,
            container_id,
        }
    }
}

pub enum AppEvent {
    /// Initial list of containers when app starts for a specific host
    InitialContainerList(HostId, Vec<Container>),
    /// A new container was created/started (host_id is in the Container)
    ContainerCreated(Container),
    /// A container was stopped/destroyed on a specific host
    ContainerDestroyed(ContainerKey),
    /// Stats update for an existing container on a specific host
    ContainerStat(ContainerKey, ContainerStats),
    /// User requested to quit
    Quit,
    /// Terminal was resized
    Resize,
    /// Move selection up
    SelectPrevious,
    /// Move selection down
    SelectNext,
    /// User pressed Enter key
    EnterPressed,
    /// User pressed Escape to exit log view
    ExitLogView,
    /// User scrolled up in log view
    ScrollUp,
    /// User scrolled down in log view
    ScrollDown,
    /// New log line received from streaming logs
    LogLine(ContainerKey, LogEntry),
}

pub type EventSender = mpsc::Sender<AppEvent>;

/// Current view state of the application
#[derive(Clone, Debug, PartialEq)]
pub enum ViewState {
    /// Viewing the container list
    ContainerList,
    /// Viewing logs for a specific container
    LogView(ContainerKey),
}
