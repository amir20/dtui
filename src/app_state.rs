use ratatui::widgets::TableState;
use std::collections::HashMap;
use tokio::sync::mpsc;

use crate::docker::DockerHost;
use crate::logs::{LogEntry, stream_container_logs};
use crate::types::{AppEvent, Container, ContainerKey, ViewState};

/// Application state that manages all runtime data
pub struct AppState {
    /// All containers indexed by (host_id, container_id)
    pub containers: HashMap<ContainerKey, Container>,
    /// Pre-sorted list of container keys for efficient rendering
    pub sorted_container_keys: Vec<ContainerKey>,
    /// Whether the application should quit
    pub should_quit: bool,
    /// Table selection state
    pub table_state: TableState,
    /// Current view (container list or log view)
    pub view_state: ViewState,
    /// Current container logs (only one container at a time)
    pub current_logs: Option<(ContainerKey, Vec<LogEntry>)>,
    /// Current scroll position (number of lines scrolled from top)
    pub log_scroll_offset: usize,
    /// Whether the user is at the bottom of the logs (for auto-scroll behavior)
    pub is_at_bottom: bool,
    /// Handle to the currently running log stream task
    pub log_stream_handle: Option<tokio::task::JoinHandle<()>>,
    /// Connected Docker hosts for log streaming
    pub connected_hosts: HashMap<String, DockerHost>,
    /// Event sender for spawning log streams
    pub event_tx: mpsc::Sender<AppEvent>,
}

impl AppState {
    /// Creates a new AppState instance
    pub fn new(
        connected_hosts: HashMap<String, DockerHost>,
        event_tx: mpsc::Sender<AppEvent>,
    ) -> Self {
        Self {
            containers: HashMap::new(),
            sorted_container_keys: Vec::new(),
            should_quit: false,
            table_state: TableState::default(),
            view_state: ViewState::ContainerList,
            current_logs: None,
            log_scroll_offset: 0,
            is_at_bottom: true,
            log_stream_handle: None,
            connected_hosts,
            event_tx,
        }
    }

    /// Processes a single event and returns whether UI should be redrawn
    pub fn handle_event(&mut self, event: AppEvent) -> bool {
        match event {
            AppEvent::InitialContainerList(host_id, container_list) => {
                self.handle_initial_container_list(host_id, container_list)
            }
            AppEvent::ContainerCreated(container) => self.handle_container_created(container),
            AppEvent::ContainerDestroyed(key) => self.handle_container_destroyed(key),
            AppEvent::ContainerStat(key, stats) => self.handle_container_stat(key, stats),
            AppEvent::Resize => true, // Always redraw on resize
            AppEvent::Quit => {
                self.should_quit = true;
                false
            }
            AppEvent::SelectPrevious => self.handle_select_previous(),
            AppEvent::SelectNext => self.handle_select_next(),
            AppEvent::EnterPressed => self.handle_enter_pressed(),
            AppEvent::ExitLogView => self.handle_exit_log_view(),
            AppEvent::ScrollUp => self.handle_scroll_up(),
            AppEvent::ScrollDown => self.handle_scroll_down(),
            AppEvent::LogLine(key, log_line) => self.handle_log_line(key, log_line),
        }
    }

    fn handle_initial_container_list(
        &mut self,
        host_id: String,
        container_list: Vec<Container>,
    ) -> bool {
        for container in container_list {
            let key = ContainerKey::new(host_id.clone(), container.id.clone());
            self.containers.insert(key.clone(), container);
            self.sorted_container_keys.push(key);
        }

        // Sort by host_id first, then by container name
        self.sorted_container_keys.sort_by(|a, b| {
            let container_a = self.containers.get(a).unwrap();
            let container_b = self.containers.get(b).unwrap();

            match container_a.host_id.cmp(&container_b.host_id) {
                std::cmp::Ordering::Equal => container_a.name.cmp(&container_b.name),
                other => other,
            }
        });

        // Select first row if we have containers
        if !self.containers.is_empty() {
            self.table_state.select(Some(0));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_created(&mut self, container: Container) -> bool {
        let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
        let sort_key = (container.host_id.clone(), container.name.clone());
        self.containers.insert(key.clone(), container);

        // Insert into sorted position
        let insert_pos = self
            .sorted_container_keys
            .binary_search_by(|probe_key| {
                let probe_container = self.containers.get(probe_key).unwrap();
                let probe_sort_key = (&probe_container.host_id, &probe_container.name);
                probe_sort_key.cmp(&(&sort_key.0, &sort_key.1))
            })
            .unwrap_or_else(|pos| pos);
        self.sorted_container_keys.insert(insert_pos, key);

        // Select first row if this is the first container
        if self.containers.len() == 1 {
            self.table_state.select(Some(0));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_destroyed(&mut self, key: ContainerKey) -> bool {
        self.containers.remove(&key);
        self.sorted_container_keys.retain(|k| k != &key);

        // Adjust selection if needed
        let container_count = self.containers.len();
        if container_count == 0 {
            self.table_state.select(None);
        } else if let Some(selected) = self.table_state.selected()
            && selected >= container_count
        {
            self.table_state.select(Some(container_count - 1));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_stat(
        &mut self,
        key: ContainerKey,
        stats: crate::types::ContainerStats,
    ) -> bool {
        if let Some(container) = self.containers.get_mut(&key) {
            container.stats = stats;
        }
        false // No force draw - just stats update
    }

    fn handle_select_previous(&mut self) -> bool {
        let container_count = self.containers.len();
        if container_count > 0 {
            let selected = self.table_state.selected().unwrap_or(0);
            if selected > 0 {
                self.table_state.select(Some(selected - 1));
            }
        }
        true // Force draw - selection changed
    }

    fn handle_select_next(&mut self) -> bool {
        let container_count = self.containers.len();
        if container_count > 0 {
            let selected = self.table_state.selected().unwrap_or(0);
            if selected < container_count - 1 {
                self.table_state.select(Some(selected + 1));
            }
        }
        true // Force draw - selection changed
    }

    fn handle_enter_pressed(&mut self) -> bool {
        // Only handle Enter in ContainerList view
        if self.view_state != ViewState::ContainerList {
            return false;
        }

        // Get the selected container
        let Some(selected_idx) = self.table_state.selected() else {
            return false;
        };

        let Some(container_key) = self.sorted_container_keys.get(selected_idx) else {
            return false;
        };

        // Switch to log view
        self.view_state = ViewState::LogView(container_key.clone());

        // Initialize log storage for this container (clear any previous logs)
        self.current_logs = Some((container_key.clone(), Vec::new()));

        // Reset scroll state - start at bottom
        self.log_scroll_offset = 0;
        self.is_at_bottom = true;

        // Stop any existing log stream
        if let Some(handle) = self.log_stream_handle.take() {
            handle.abort();
        }

        // Start streaming logs for this container
        if let Some(host) = self.connected_hosts.get(&container_key.host_id) {
            let host_clone = host.clone();
            let container_id = container_key.container_id.clone();
            let tx_clone = self.event_tx.clone();

            let handle = tokio::spawn(async move {
                stream_container_logs(host_clone, container_id, tx_clone).await;
            });

            self.log_stream_handle = Some(handle);
        }

        true // Force draw - view changed
    }

    fn handle_exit_log_view(&mut self) -> bool {
        // Only handle Escape when in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Stop log streaming
        if let Some(handle) = self.log_stream_handle.take() {
            handle.abort();
        }

        // Clear current logs
        self.current_logs = None;

        // Switch back to container list view
        self.view_state = ViewState::ContainerList;

        true // Force draw - view changed
    }

    fn handle_scroll_up(&mut self) -> bool {
        // Only handle scroll in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Scroll up (decrease offset)
        if self.log_scroll_offset > 0 {
            self.log_scroll_offset = self.log_scroll_offset.saturating_sub(1);
            self.is_at_bottom = false; // User scrolled away from bottom
            return true; // Force draw
        }

        false
    }

    fn handle_scroll_down(&mut self) -> bool {
        // Only handle scroll in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Only scroll if we have logs
        if self.current_logs.is_some() {
            // Increment scroll offset
            self.log_scroll_offset = self.log_scroll_offset.saturating_add(1);

            // Will be clamped in UI and is_at_bottom will be recalculated there
            return true; // Force draw
        }

        false
    }

    fn handle_log_line(&mut self, key: ContainerKey, log_entry: LogEntry) -> bool {
        // Only add log line if we're currently viewing this container's logs
        if let Some((current_key, logs)) = &mut self.current_logs
            && current_key == &key
        {
            logs.push(log_entry);

            // Only auto-scroll if user is at the bottom
            if self.is_at_bottom {
                // Scroll will be updated to show bottom in UI
            }

            return true; // Force draw - new log line for currently viewed container
        }

        // Ignore log lines for containers we're not viewing
        false
    }
}
