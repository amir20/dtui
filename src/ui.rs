use ratatui::{
    Frame,
    layout::Constraint,
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Cell, Paragraph, Row, Table, TableState, Wrap},
};
use std::collections::HashMap;

use crate::app_state::AppState;
use crate::types::{Container, ContainerKey, ViewState};

/// Pre-allocated styles to avoid recreation every frame
pub struct UiStyles {
    pub high: Style,
    pub medium: Style,
    pub low: Style,
    pub header: Style,
    pub border: Style,
    pub selected: Style,
    pub timestamp: Style,
}

impl Default for UiStyles {
    fn default() -> Self {
        Self {
            high: Style::default().fg(Color::Red),
            medium: Style::default().fg(Color::Yellow),
            low: Style::default().fg(Color::Green),
            header: Style::default()
                .fg(Color::Cyan)
                .add_modifier(Modifier::BOLD),
            border: Style::default().fg(Color::White),
            selected: Style::default()
                .bg(Color::DarkGray)
                .add_modifier(Modifier::BOLD),
            timestamp: Style::default()
                .fg(Color::Yellow)
                .add_modifier(Modifier::BOLD),
        }
    }
}

/// Renders the main UI - either container list or log view
pub fn render_ui(f: &mut Frame, state: &mut AppState, styles: &UiStyles) {
    match &state.view_state {
        ViewState::ContainerList => {
            // Calculate unique hosts to determine if host column should be shown
            let unique_hosts: std::collections::HashSet<_> =
                state.containers.keys().map(|key| &key.host_id).collect();
            let show_host_column = unique_hosts.len() > 1;

            render_container_list(
                f,
                &state.containers,
                &state.sorted_container_keys,
                styles,
                &mut state.table_state,
                show_host_column,
            );
        }
        ViewState::LogView(container_key) => {
            let container_key = container_key.clone();
            render_log_view(f, &container_key, state, styles);
        }
    }
}

/// Renders the container list view
fn render_container_list(
    f: &mut Frame,
    containers: &HashMap<ContainerKey, Container>,
    sorted_container_keys: &[ContainerKey],
    styles: &UiStyles,
    table_state: &mut TableState,
    show_host_column: bool,
) {
    let size = f.area();

    // Use pre-sorted list instead of sorting every frame
    let rows: Vec<Row> = sorted_container_keys
        .iter()
        .filter_map(|key| containers.get(key))
        .map(|c| create_container_row(c, styles, show_host_column))
        .collect();

    let header = create_header_row(styles, show_host_column);
    let table = create_table(rows, header, containers.len(), styles, show_host_column);

    f.render_stateful_widget(table, size, table_state);
}

/// Renders the log view for a specific container
fn render_log_view(
    f: &mut Frame,
    container_key: &ContainerKey,
    state: &mut AppState,
    styles: &UiStyles,
) {
    let size = f.area();

    // Get container info
    let container_name = state
        .containers
        .get(container_key)
        .map(|c| c.name.as_str())
        .unwrap_or("Unknown");

    // Get logs for this container (only if it matches current_logs)
    let (log_lines, num_lines) = if let Some((key, logs)) = &state.current_logs {
        if key == container_key {
            // Format log entries into styled lines with colored timestamps
            let lines: Vec<Line> = logs
                .iter()
                .map(|entry| {
                    let timestamp_str = entry.timestamp.format("%Y-%m-%d %H:%M:%S").to_string();
                    Line::from(vec![
                        Span::styled(timestamp_str, styles.timestamp),
                        Span::raw(" "),
                        Span::raw(entry.message.clone()),
                    ])
                })
                .collect();
            let num_lines = lines.len();
            (lines, num_lines)
        } else {
            (Vec::new(), 0)
        }
    } else {
        (Vec::new(), 0)
    };

    // Calculate visible height (subtract 2 for borders)
    let visible_height = size.height.saturating_sub(2) as usize;

    // Calculate max scroll position
    let max_scroll = if num_lines > visible_height {
        num_lines.saturating_sub(visible_height)
    } else {
        0
    };

    // Determine actual scroll offset
    let actual_scroll = if state.is_at_bottom {
        // Auto-scroll to bottom
        max_scroll
    } else {
        // Use manual scroll position, but clamp to max
        state.log_scroll_offset.min(max_scroll)
    };

    // Update is_at_bottom based on actual position
    state.is_at_bottom = actual_scroll >= max_scroll;

    // Update scroll offset to actual (for proper clamping)
    state.log_scroll_offset = actual_scroll;

    // Create log widget with scrolling
    let log_widget = Paragraph::new(log_lines)
        .block(
            Block::default()
                .borders(Borders::ALL)
                .title(format!(
                    "Logs: {} ({}) - Press ESC to return {}",
                    container_name,
                    container_key.host_id,
                    if state.is_at_bottom {
                        "[AUTO]"
                    } else {
                        "[MANUAL]"
                    }
                ))
                .style(styles.border),
        )
        .wrap(Wrap { trim: false })
        .scroll((actual_scroll as u16, 0));

    f.render_widget(log_widget, size);
}

/// Creates a table row for a single container
fn create_container_row<'a>(
    container: &'a Container,
    styles: &UiStyles,
    show_host_column: bool,
) -> Row<'a> {
    let cpu_bar = create_progress_bar(container.stats.cpu, 20);
    let cpu_style = get_percentage_style(container.stats.cpu, styles);

    let memory_bar = create_progress_bar(container.stats.memory, 20);
    let memory_style = get_percentage_style(container.stats.memory, styles);

    let network_tx = format_bytes_per_sec(container.stats.network_tx_bytes_per_sec);
    let network_rx = format_bytes_per_sec(container.stats.network_rx_bytes_per_sec);

    let mut cells = vec![
        Cell::from(container.id.as_str()),
        Cell::from(container.name.as_str()),
    ];

    if show_host_column {
        cells.push(Cell::from(container.host_id.as_str()));
    }

    cells.extend(vec![
        Cell::from(cpu_bar).style(cpu_style),
        Cell::from(memory_bar).style(memory_style),
        Cell::from(network_tx),
        Cell::from(network_rx),
        Cell::from(container.status.as_str()),
    ]);

    Row::new(cells)
}

/// Creates a text-based progress bar with percentage
fn create_progress_bar(percentage: f64, width: usize) -> String {
    let percentage = percentage.clamp(0.0, 100.0);
    let filled_width = ((percentage / 100.0) * width as f64).round() as usize;
    let empty_width = width.saturating_sub(filled_width);

    let bar = format!("{}{}", "█".repeat(filled_width), "░".repeat(empty_width));

    format!("{} {:5.1}%", bar, percentage)
}

/// Formats bytes per second into a human-readable string (KB/s, MB/s, GB/s)
fn format_bytes_per_sec(bytes_per_sec: f64) -> String {
    const KB: f64 = 1024.0;
    const MB: f64 = KB * 1024.0;
    const GB: f64 = MB * 1024.0;

    if bytes_per_sec >= GB {
        format!("{:.2}GB/s", bytes_per_sec / GB)
    } else if bytes_per_sec >= MB {
        format!("{:.2}MB/s", bytes_per_sec / MB)
    } else if bytes_per_sec >= KB {
        format!("{:.1}KB/s", bytes_per_sec / KB)
    } else {
        format!("{:.0}B/s", bytes_per_sec)
    }
}

/// Returns the appropriate style based on percentage value
pub fn get_percentage_style(value: f64, styles: &UiStyles) -> Style {
    if value > 80.0 {
        styles.high
    } else if value > 50.0 {
        styles.medium
    } else {
        styles.low
    }
}

/// Creates the table header row
fn create_header_row(styles: &UiStyles, show_host_column: bool) -> Row<'static> {
    let mut headers = vec!["ID", "Name"];

    if show_host_column {
        headers.push("Host");
    }

    headers.extend(vec!["CPU %", "Memory %", "Net TX", "Net RX", "Status"]);

    Row::new(headers).style(styles.header).bottom_margin(1)
}

/// Creates the complete table widget
fn create_table<'a>(
    rows: Vec<Row<'a>>,
    header: Row<'static>,
    container_count: usize,
    styles: &UiStyles,
    show_host_column: bool,
) -> Table<'a> {
    let mut constraints = vec![
        Constraint::Length(12), // Container ID
        Constraint::Fill(1),    // Name (flexible)
    ];

    if show_host_column {
        constraints.push(Constraint::Length(20)); // Host
    }

    constraints.extend(vec![
        Constraint::Length(28), // CPU progress bar (20 chars + " 100.0%")
        Constraint::Length(28), // Memory progress bar (20 chars + " 100.0%")
        Constraint::Length(12), // Network TX (1.23MB/s)
        Constraint::Length(12), // Network RX (4.56MB/s)
        Constraint::Length(15), // Status
    ]);

    Table::new(rows, constraints)
        .header(header)
        .block(
            Block::default()
                .borders(Borders::ALL)
                .title(format!(
                    "Docker Container CPU Monitor - {} containers (↑/↓ to navigate, 'q' to quit)",
                    container_count
                ))
                .style(styles.border),
        )
        .row_highlight_style(styles.selected)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_percentage_style_thresholds() {
        let styles = UiStyles::default();

        // Test low threshold (green)
        let low_style = get_percentage_style(30.0, &styles);
        assert_eq!(low_style.fg, Some(Color::Green));

        // Test medium threshold (yellow)
        let medium_style = get_percentage_style(65.0, &styles);
        assert_eq!(medium_style.fg, Some(Color::Yellow));

        // Test high threshold (red)
        let high_style = get_percentage_style(85.0, &styles);
        assert_eq!(high_style.fg, Some(Color::Red));

        // Test boundary cases
        assert_eq!(get_percentage_style(50.0, &styles).fg, Some(Color::Green));
        assert_eq!(get_percentage_style(50.1, &styles).fg, Some(Color::Yellow));
        assert_eq!(get_percentage_style(80.0, &styles).fg, Some(Color::Yellow));
        assert_eq!(get_percentage_style(80.1, &styles).fg, Some(Color::Red));
    }
    #[test]
    fn test_color_coding_boundaries() {
        let styles = UiStyles::default();

        // Test exact boundary values
        assert_eq!(
            get_percentage_style(0.0, &styles).fg,
            Some(Color::Green),
            "0% should be green"
        );
        assert_eq!(
            get_percentage_style(50.0, &styles).fg,
            Some(Color::Green),
            "50% should be green"
        );
        assert_eq!(
            get_percentage_style(50.1, &styles).fg,
            Some(Color::Yellow),
            "50.1% should be yellow"
        );
        assert_eq!(
            get_percentage_style(80.0, &styles).fg,
            Some(Color::Yellow),
            "80% should be yellow"
        );
        assert_eq!(
            get_percentage_style(80.1, &styles).fg,
            Some(Color::Red),
            "80.1% should be red"
        );
        assert_eq!(
            get_percentage_style(100.0, &styles).fg,
            Some(Color::Red),
            "100% should be red"
        );
    }
}
