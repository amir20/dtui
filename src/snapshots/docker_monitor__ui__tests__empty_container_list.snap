---
source: src/ui.rs
assertion_line: 430
expression: "format!(\"{:?}\", buffer)"
---
Buffer {
    area: Rect { x: 0, y: 0, width: 80, height: 20 },
    content: [
        "┌Docker Container CPU Monitor - 0 containers (↑/↓ to navigate, 'q' to quit)────┐",
        "│  Host         CPU %        Memory %     Net TX      Net RX       Status      │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "│                                                                              │",
        "└──────────────────────────────────────────────────────────────────────────────┘",
    ],
    styles: [
        x: 0, y: 0, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 1, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 79, y: 1, fg: White, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
